src/proj.rs

```diff
@@ -0,0 +1,791 @@
+        let t = &self.transform;
+        let tx = t[2];
+        let ty = t[5];
+
+        (x + tx,
+         y + ty)
+    }
+}
+
+/// Combines two projective transformations
+/// (matrix multiplication)
+impl Mul<Proj> for Proj {
+    type Output = Proj;
+
+    fn mul(self, rhs: Proj) -> Proj {
+        use TransformationClass as TC;
+        let t = mul3x3(self.transform, rhs.transform);
+        let i = mul3x3(rhs.inverse, self.inverse);
+
+        let class = match (self.class, rhs.class) {
+            (TC::Translation, TC::Translation) => TC::Translation,
+            (TC::Translation, TC::Affine) => TC::Affine,
+            (TC::Affine, TC::Translation) => TC::Affine,
+            (TC::Affine, TC::Affine)      => TC::Affine,
+            (_, _) => TC::Projection
+        };
+
+        Proj { transform: t, inverse: i, class: class }
+    }
+}
+
+/// Combines two projective transformations
+/// (matrix multiplication)
+impl<'a, 'b> Mul<&'b Proj> for &'a Proj {
+    type Output = Proj;
+
+    fn mul(self, rhs: &Proj) -> Proj {
+        use TransformationClass as TC;
+        let t = mul3x3(self.transform, rhs.transform);
+        let i = mul3x3(rhs.inverse, self.inverse);
+        let class = match (self.class, rhs.class) {
+            (TC::Translation, TC::Translation) => TC::Translation,
+            (TC::Translation, TC::Affine) => TC::Affine,
+            (TC::Affine, TC::Translation) => TC::Affine,
+            (_, _) => TC::Projection
+        };
+
+        Proj { transform: t, inverse: i, class: class }
+    }
+}
+
+impl Mul<(f32, f32)> for Proj {
+    type Output = (f32, f32);
+
+    fn mul(self, rhs: (f32, f32)) -> (f32, f32) {
+        let (x,y) = rhs;
+        self.warp_p(x, y)
+    }
+}
+
+impl<'a, 'b> Mul<&'b (f32, f32)> for &'a Proj {
+    type Output = (f32, f32);
+
+    fn mul(self, rhs: &(f32, f32)) -> (f32, f32) {
+        let (x,y) = *rhs;
+        self.warp_p(x, y)
+    }
+}
+
+/// Performs projective transformation to an image. Allocates an output image with same
+/// dimensions as the input image. Output pixels outside the input image are set to `default`.
+/// Proj `homography` defines a mapping from coordinates in the `output` image to coordinates of
+/// the `image`.
+pub fn warp_new<P>(homography: &Proj,
+                   image: &Image<P>,
+                   interpolation: Interpolation,
+                   default: P
+                  ) -> Image<P> 
+where
+P: Pixel + Send + Sync + 'static,
+<P as Pixel>::Subpixel: Send + Sync,
+<P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let (width, height) = image.dimensions();
+    let mut out = ImageBuffer::new(width, height);
+    warp(homography, image, interpolation, default, &mut out);
+
+    out
+}
+
+/// Performs projective transformation `homography` mapping pixels from
+/// `image` into `&mut output`.
+/// Proj `homograpy` defines coordinate mapping from `out` to `image`.
+pub fn warp<P>(homography: &Proj,
```

Comment:
In the rest of this library we use `my_func` to name functions which allocate a new return value, and `my_func_mut` to refer to those that mutate an image in place. There aren't yet any any public functions that take a preallocated output, so we don't have a naming convention for this.

I suggest

`warp(...) -> Image`
`warp_into(..., &mut)`

but am open to better suggestions.

Reply:
I like warp and warp_into.